include/OpenSpaceToolkit/Astrodynamics/NumericalSolver.hpp

```diff
@@ -221,7 +228,30 @@ class NumericalSolver
         const SystemOfEquationsWrapper& aSystemOfEquations
     );
 
-    /// @brief              Perform numerical integration for a certain duration
+    /// @brief              Perform numerical integration from a start time till a condition is met, or end time is
+    /// reached
+    ///
+    /// @code
+    ///                     StateVector stateVector = numericalSolver.integrateTime(stateVector,
+    ///                     aStartTime, anEndTime, SystemOfEquations, anEventCondition) ;
+    /// @endcode
+    /// @param              [in] anInitialStateVector An initial n-dimensional state vector to begin integrating at
+    /// @param              [in] aStartTime A time to begin integrating from
+    /// @param              [in] anEndTime A maximum time to to integrate to
+    /// @param              [in] aSystemOfEquations An std::function wrapper with a particular signature that
+    ///                     boost::odeint accepts to perform numerical integration
+    /// @param              [in] anEventCondition An event condition
+    /// @return             Solution
+
+    Solution integrateTime(
+        const StateVector& anInitialStateVector,
+        const Real& aStartTime,
+        const Real& anEndTime,
+        const SystemOfEquationsWrapper& aSystemOfEquations,
+        const Shared<EventCondition>& anEventCondition
```

Comment:
What's the reason for passing this in as a Shared pointer instead of plain old passing by reference? Do we need it to be allocated on the heap for some reason? This seems like a user level thing and having to wrap your inputs in a shared pointer can be a little cumbersome IMO.

Reply:
Since `EventCondition` is an abstract class, we have to pass it via a shared pointer.
The flow is:
Base Abstract `EventCondition` class
|_> Child classes that implement the various conditions via the evaluate method
     |_> evaluate

Therefore for the polymorphism this has to be a shared pointer.

Reply:
Maybe I'm misunderstanding the scope you're talking about, but I was a little skeptical because that seems like a decent restriction on polymorphism.
So I tried it out:
https://github.com/open-space-collective/open-space-toolkit-astrodynamics/pull/155

Changing it to a non-pointer type seems to work with at least one of the tests.

Reply:
Kyle was right! We can pass this via reference as we are only using the parent methods so there are no issues with object slicing.